docs/site/Deploying-to-IBM-Cloud.md

```diff
@@ -0,0 +1,211 @@
+---
+lang: en
+title: 'Deploying to IBM Cloud'
+keywords: LoopBack
+sidebar: lb4_sidebar
+permalink: /doc/en/lb4/Deploying-to-IBM-Cloud.html
+---
+
+This is a guide for deploying LoopBack 4 (LB4) apps to IBM Cloud. In the setup
+explained below, your app will use a local instance of Cloundant when running
+locally, and a provisioned Cloudant service when running on the cloud.
+
+Before we begin make sure the following are installed.
+
+1.  [Node.js](https://nodejs.org/en/download/) >= 8.9.0.
+2.  [LB4 CLI](https://www.npmjs.com/package/@loopback/cli) (`lb4` command).
+3.  [Cloud Foundy CLI](https://docs.cloudfoundry.org/cf-cli/install-go-cli.html).
+4.  [Docker](https://www.docker.com/).
+
+Consider the
+["todo" example](https://github.com/strongloop/loopback-next/tree/master/examples/todo)
+from the [loopback-next repository](https://github.com/strongloop/loopback-next)
+as a basis for the instruction.
+
+You can quickly clone the "todo" example app by running the command:
+
+```sh
+$ lb4 example todo
+```
+
+## Local setup
+
+### Database
+
+1.  [Download Cloudant image](https://hub.docker.com/r/ibmcom/cloudant-developer/).
+2.  Ensure port 8080 is free and start Cloudant:
+
+```
+docker run \
+      --volume cloudant:/srv \
+      --name cloudant-developer \
+      --publish 8080:80 \
+      --hostname cloudant.dev \
+      ibmcom/cloudant-developer
+```
+
+3.  Log on to the local Cloudant Dashboard using `admin`:`pass` as the
+    crendentials.
+4.  Create a database named `todo`.
+
+### Updates to the app
+
+#### Updating datasource
+
+Update `db.datasource.json` to use the Cloudant connector.
+
+```js
+{
+  "name": "db",
+  "connector": "cloudant",
+  "url": "http://admin:pass@localhost:8080",
+  "database": "todo",
+  "modelIndex": ""
+}
+```
+
+Install the `loopback-connector-cloudant` package.
+
+```sh
+$ npm i loopback-connector-cloudant -s
+```
+
+#### Updating npm script
+
+Remove the `prestart` script from `package.json`, since we don't want to do any
```

Comment:
I'm not sure I agree with this part! Building locally and deploying (without the `node_modules`) folder seems weird. What if there is a new dependency released vs. the target your app was built with? Not to mention it seems like a pain to have to build the app yourself before deploying it when the build folder isn't stored in GitHub. 

The way I personally see builds working is as follows: 
- BASIC: Build the app on the cloud as you would locally. 
- Intermediate: Set up a CI/CD pipeline to build / deploy the app, something like Travis / IBM Cloud DevOps offering. This would help limit downtime during a switch if not using blue-green deployments.

Building locally should not be a thing and the deploy should be with the dependencies it's built with IMO.

Reply:
Looks like building on IBM Cloud will require `lb-tsc` to be installed globally. Moving `@loopback/build` to `dependencies` does not make `lb-tsc` available to npm script like it was installed globally.

Any suggestions?

Reply:
Really? Because I don't have typescript installed globally and have never had issues with `lb-tsc` not working locally. @raymondfeng any ideas as to why this might happen? Only thing that comes to my mind is it might be related to Node 8?

Reply:
Maybe something to do with the environment on IBM Cloud. I tried logging in to the shell, there's no `node` or `npm`.

Reply:
> Building locally and deploying (without the node_modules) folder seems weird. What if there is a new dependency released vs. the target your app was built with? 

Or what if the public npm registry goes down and our production box cannot install the dependencies?

IMO, if we deploy from a local computer (instead of using a build/deploy pipeline), then we should run the build (TypeScript compilation) locally AND also deploy `node_modules` dependencies as we have installed them locally.

At least this was our conclusion when we were discussing different deployment options back in 2014/2015. We created https://github.com/strongloop/strong-build to make that process easy to automate.

@rmg could you please chime in? What's your opinion on the deployment process for Node.js apps requiring a build step?